### ISSUE TITLE
Views not released after modal ViewController dismissed.

### ISSUE THREAD

**Issue by n-studio** [colinta/teacup#88]
Hi!

I'm using Teacup for the first time and I noticed that in several situations my views are not released when the ViewController is.
For instance, when I dismiss a modal ViewController, all the views stay alive.

I made a sample app to illustrate the issue:
https://github.com/n-studio/TeacupMemoryLeak

There is a with-Teacup and a without-Teacup version. The without-Teacup version works perfectly, whereas in the with-Teacup version Instruments shows me still living UIViews.

Am I doing something wrong?

**Comment by colinta** [comment]
Interesting, I've never seen this! I'll def check it out. Thanks!

**Comment by goodwill** [comment]
I think I come across similar behaviour. Anything that use teacup stylesheet the whole controller would not dealloc at all, please fix that ASAP- this is introducing major memory leak to the app.

**Comment by goodwill** [comment]
Here is another demo. The dealloc on controller is never called- if you take away the stylesheet and layout then memory will release properly.

https://github.com/goodwill/teacup-leak-hello2

**Comment by goodwill** [comment]
Going further into the problem, I realise UIView inside the UIViewController is always not dealloc'ed if a layout do block exist. The latest version allow the UIViewController to release, but not the UIView. Constructing the UI using traditional viewDidLoad approach does not cause such problem.

I use this block of code to trigger logging to console when a UIView's dealloc is called:

``` ruby
UIView.class_eval do
  def dealloc
    ap "#{self.class.name} dealloc"
    super
  end
end
```

**Comment by goodwill** [comment]
I did another project to show the leak:

https://github.com/goodwill/teacup-leaks

Don't know why its so easy to leak with teacup? Just adding layout(:style) do .. end then all the table view cell are not released. I think there are some sort of cyclic references in the code need to be simplified and refactored to avoid such leakage.

**Comment by bogardon** [comment]
any progress on this issue? it' s pretty serious...:(

**Comment by colinta** [comment]
Actually, I've been talking with Laurent about it. There are similar issues in BubbleWrap, we're trying to come up with a good solution to this one.  No ETA, though.  It's not as easy as cyclic references (though I'm not counting those out, either), it has something to do with blocks retaining `self`.

**Comment by bogardon** [comment]
thanks for the update @colinta.

i did notice that even a clean rubymotion project with nothing implemented in `appDidFinishLaunching` has tons of leaks. really hope Rubymotion resolves these soon D:

**Comment by goodwill** [comment]
I have ended up extracted the style sheet feature of teacup in a separate gem to keep the application simple (thus avoid memory leak). Until we have a good way to avoid such leak, teacup is very good library but with a pretty big side effect on memory management that might make production use challenging.

```
—
```

**Comment by ghost** [comment]
I'm getting an identical problem with our app. When our view controller gets popped from the UINavigationController it gets released, but as soon as I put
layout(view) do
end
into viewDidLoad it's no longer deallocated when popped.

Goodwill, I saw your fork and your commit https://github.com/goodwill/teacup/commit/d5bf4c050d9f4026827b6487cd69cee58499f20e. Unfortunately, that doesn't fix the problem. Does it work for you? What about the work you did factoring out the stylesheet aspect of teacup to a separate gem? Are you talking about your fork or something else?

Are there any other updates about this? Our app loads lots of images per view controller, and so this stuff not getting released is going to be problematic for us.

**Comment by goodwill** [comment]
Because we prioritise fixing memory leak instead of all the great features
(aka this thing is not going to replace teacup, its a lot more handicapped
put it this way, but it provides basic stylesheet feature, with
motion-layout combined it does most what we need)

https://github.com/apriltg/plastic_cup

Until we have a great memory leak free solution this might be a good
compromise. We made things simple (e.g. avoid instance_eval on blocks which
seems suspect to memory leak), try to keep as least instance reference as
possible.

On Wed, Sep 18, 2013 at 5:11 AM, paulcarpenterSE
notifications@github.comwrote:

> I'm getting an identical problem with our app. When our view controller
> gets popped from the UINavigationController it gets released, but as soon
> as I put
> layout(view) do
> end
> into viewDidLoad it's no longer deallocated when popped.
> 
> Goodwill, I saw your fork and your commit goodwill@d5bf4c0https://github.com/goodwill/teacup/commit/d5bf4c050d9f4026827b6487cd69cee58499f20e.
> Unfortunately, that doesn't fix the problem. Does it work for you? What
> about the work you did factoring out the stylesheet aspect of teacup to a
> separate gem? Are you talking about your fork or something else?
> 
> Are there any other updates about this? Our app loads lots of images per
> view controller, and so this stuff not getting released is going to be
> problematic for us.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/rubymotion/teacup/issues/88#issuecomment-24622720
> .

## 

Regards,

William Yeung

Consultant

jquery+rubyonrails

Tofugear Limited

E: william@tofugear.com
T: +852 21519830
M: +852 92538475
W: http://tofugear.com

**Comment by colinta** [comment]
Very cool, but why not put these efforts towards improving teacup?

**Comment by goodwill** [comment]
We tried, unfortunately it proves to be too hard for our team. We are significantly trimming down the magic to avoid memory leak. Not like we actually patched the source.

—

**Comment by ghost** [comment]
goodwill, thanks for linking me to your gem and putting it up here.

**Comment by colinta** [comment]
Resolved (?) in 2.1.12.

**Comment by n-studio** [comment]
Unfortunately the leak seems to be still here....

https://github.com/n-studio/TeacupMemoryLeak
